File: ex3_capitalize.rb
=begin
## Exercício 3: Capitalizar Nomes

- **Descrição**: Crie uma função que recebe um array de nomes e retorna um novo array com todos os nomes capitalizados (primeira letra maiúscula).

- **Regra**: Considere nomes compostos.

- **Input**: ["ana maria", "joão silva"]
- **Output**: ["Ana Maria", "João Silva"]
- **Input**: ["pedro"]
- **Output**: ["Pedro"]
=end


def capitalize_names(names)
  capitalized_names = []

  names.each do |name|
    capitalized_name = ""
    capitalize_next = true

    name.chars.each do |char|
      if capitalize_next && char.match?(/[a-z]/) #Usando regex para verificar se o caractere está minusculo e então capitalizar.
        capitalized_name << char.upcase #Capitalizando o caractere
        capitalize_next = false #Setando para false para evitar que o próximo caractere seja capitalizado
      else
        capitalized_name << char #Se a condição acima não for atendida, o caractere é adicinado a variavel sem nenhuma alteração
      end

      capitalize_next = true if char == " " #Se o caractere atual que será iterado for um espaço, ele seta a variavel para true indicando que o proximo devera ser capitalizado
    end

    capitalized_names << capitalized_name #Quando todos caracteres forem processados, ele são inseridos no array capitalized_names
  end

  capitalized_names #Retorna o array
end

# Test cases
input1 = ["ana maria", "joão silva"]
input2 = ["pedro"]

puts "Input: #{input1}"
puts "Output: #{capitalize_names(input1)}"

puts "Input: #{input2}"
puts "Output: #{capitalize_names(input2)}"

File: ex3_capitalize.go
package main

import (
	"fmt"
	"strings"
)

// Função capitalizeName recebe um nome como entrada e retorna o nome capitalizado
func capitalizeName(name string) string {
	return strings.Title(name)
}

// Função capitalizeNames recebe um array de nomes como entrada e retorna um novo array com todos os nomes capitalizados
func capitalizeNames(names []string) []string {
	// Cria um novo array para armazenar os nomes capitalizados, com o mesmo tamanho do array de entrada
	capitalizedNames := make([]string, len(names))

	// Percorre cada nome no array de entrada
	for i, name := range names {
		// Capitaliza o nome usando a função capitalizeName e armazena-o no novo array
		capitalizedNames[i] = capitalizeName(name)
	}

	// Retorna o novo array com os nomes capitalizados
	return capitalizedNames
}

func main() {
	// Definindo os inputs de exemplo
	input1 := []string{"ana maria", "joão silva"}
	input2 := []string{"pedro"}

	// Capitaliza os nomes dos inputs usando a função capitalizeNames
	output1 := capitalizeNames(input1)
	output2 := capitalizeNames(input2)

	// Imprime os inputs e outputs
	fmt.Println("Input:", input1)
	fmt.Println("Output:", output1)
	fmt.Println("Input:", input2)
	fmt.Println("Output:", output2)
}
